diff --git a/arch/riscv/kernel/setup.c b/arch/riscv/kernel/setup.c
index 376d2827e736..ae70baf857c0 100644
--- a/arch/riscv/kernel/setup.c
+++ b/arch/riscv/kernel/setup.c
@@ -303,6 +303,12 @@ void __init setup_arch(char **cmdline_p)
 	if (IS_ENABLED(CONFIG_RISCV_ISA_ZICBOM) &&
 	    riscv_isa_extension_available(NULL, ZICBOM))
 		riscv_noncoherent_supported();
+
+	// ask the hypervisor to preload all pages
+	// TODO: read the total memory size and expose it to the hypervisor with the call
+	sbi_ecall(0x509999, 0, 0, 0, 0, 0, 0, 0);
+	// ask the security monitor to promote the VM to a confidential VM
+	sbi_ecall(0x510000, 1000, 0, 0, 0, 0, 0, 0);		
 }
 
 static int __init topology_init(void)
