diff --git a/lib/sbi/sbi_hart.c b/lib/sbi/sbi_hart.c
index a0ab0c6..8e2e838 100644
--- a/lib/sbi/sbi_hart.c
+++ b/lib/sbi/sbi_hart.c
@@ -369,7 +369,7 @@ static int sbi_hart_smepmp_configure(struct sbi_scratch *scratch,
 	pmp_disable(SBI_SMEPMP_RESV_ENTRY);

 	/* Program M-only regions when MML is not set. */
-	pmp_idx = 0;
+	pmp_idx = 2;
 	sbi_domain_for_each_memregion(dom, reg) {
 		/* Skip reserved entry */
 		if (pmp_idx == SBI_SMEPMP_RESV_ENTRY)
@@ -395,7 +395,7 @@ static int sbi_hart_smepmp_configure(struct sbi_scratch *scratch,
 	csr_set(CSR_MSECCFG, MSECCFG_MML);

 	/* Program shared and SU-only regions */
-	pmp_idx = 0;
+	pmp_idx = 2;
 	sbi_domain_for_each_memregion(dom, reg) {
 		/* Skip reserved entry */
 		if (pmp_idx == SBI_SMEPMP_RESV_ENTRY)
@@ -432,7 +432,7 @@ static int sbi_hart_oldpmp_configure(struct sbi_scratch *scratch,
 {
 	struct sbi_domain_memregion *reg;
 	struct sbi_domain *dom = sbi_domain_thishart_ptr();
-	unsigned int pmp_idx = 0;
+	unsigned int pmp_idx = 2;
 	unsigned int pmp_flags;
 	unsigned long pmp_addr;

diff --git a/lib/sbi/sbi_hsm.c b/lib/sbi/sbi_hsm.c
index 3d60ceb..45c602d 100644
--- a/lib/sbi/sbi_hsm.c
+++ b/lib/sbi/sbi_hsm.c
@@ -26,6 +26,8 @@
 #include <sbi/sbi_timer.h>
 #include <sbi/sbi_console.h>

+extern void ace_setup_this_hart();
+
 #define __sbi_hsm_hart_change_state(hdata, oldstate, newstate)		\
 ({									\
 	long state = atomic_cmpxchg(&(hdata)->state, oldstate, newstate); \
@@ -154,6 +156,11 @@ void __noreturn sbi_hsm_hart_start_finish(struct sbi_scratch *scratch,
 	next_mode = scratch->next_mode;
 	hsm_start_ticket_release(hdata);

+	// ACE START
+	// temporal hack to reconfigure PMPs after they have been reconfigured by OpenSBI.
+	ace_setup_this_hart();
+	// ACE END
+
 	sbi_hart_switch_mode(hartid, next_arg1, next_addr, next_mode, false);
 }

